File: frontend/src/api/reports.ts
import type {
  AttendanceReport,
  EngagementReport,
  LearningOutcomeReport,
  ParticipationReport,
} from "../types/report";
import { ApiError } from "./http";

// Relative path: proxied to FastAPI in dev, same-origin in prod.
const BASE = (import.meta.env.VITE_API_BASE ?? "") + "/api/reports";

// Re-exported so callers get ApiError from the same module as the fetchers.
export { ApiError } from "./http";

/**
 * Not http.ts's error path: these routes answer "no active challenge" with a
 * structured `detail: {code, message}` (the same shape the enroll route uses),
 * and http.ts would surface that object where a string is expected. Unwrap both
 * shapes here — plain-string details still arrive from the auth guards.
 */
async function throwApiError(res: Response): Promise<never> {
  const body = await res.json().catch(() => null);
  const detail = body?.detail;
  const message =
    typeof detail === "string" ? detail : (detail?.message ?? res.statusText);
  throw new ApiError(res.status, message);
}

async function request<T>(path: string, init: RequestInit = {}): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    credentials: "include",
    headers: { "Content-Type": "application/json", ...init.headers },
    ...init,
  });
  if (!res.ok) await throwApiError(res);
  return res.json() as Promise<T>;
}

/** The server's `filename="…"`, so a download is named the same everywhere. */
function filenameFrom(disposition: string | null, fallback: string): string {
  const match = disposition?.match(/filename="([^"]+)"/);
  return match ? match[1] : fallback;
}

/**
 * `?challenge_id=N`, or nothing at all when no challenge is selected.
 *
 * Omitted rather than sent empty: the server reads a missing parameter as "the
 * challenge running right now", which is what the dashboard shows before the
 * selector has loaded and what every caller wanted before US-23 added one.
 */
function scope(challengeId?: number): string {
  return challengeId === undefined ? "" : `?challenge_id=${challengeId}`;
}

/**
 * Participation and the per-week completion funnel (FR-F1 / US-21). Defaults to
 * the campus's active challenge. Throws ApiError(404) when nothing is published,
 * or when the requested challenge is not this campus's published one.
 */
export function getParticipationReport(
  challengeId?: number,
): Promise<ParticipationReport> {
  return request<ParticipationReport>(`/participation${scope(challengeId)}`);
}

/**
 * Auto-vs-manual attendance breakdown (FR-F2 / US-22). 404s on the same condition
 * /participation does — every report route resolves its challenge the same way.
 */
export function getAttendanceReport(
  challengeId?: number,
): Promise<AttendanceReport> {
  return request<AttendanceReport>(`/attendance${scope(challengeId)}`);
}

/**
 * Content views and guide usage (FR-F3 / US-23). `challenge_id` is what US-23's
 * "both can be viewed per challenge" asks for; it is accepted by every route here
 * so the dashboard's cards always describe the same challenge.
 */
export function getEngagementReport(
  challengeId?: number,
): Promise<EngagementReport> {
  return request<EngagementReport>(`/engagement${scope(challengeId)}`);
}

/**
 * Mean assessment score per learning-outcome tag (FR-F4 / US-24). 404s on the same
 * condition the reports above do — every route here resolves its challenge the
 * same way, which is what lets the dashboard treat one 404 as "nothing published".
 *
 * US-24's Gherkin never asks for `challengeId`; it is accepted anyway, like on
 * every route here, so the dashboard's cards always describe the same challenge.
 */
export function getLearningOutcomeReport(
  challengeId?: number,
): Promise<LearningOutcomeReport> {
  return request<LearningOutcomeReport>(`/outcomes${scope(challengeId)}`);
}

/**
 * The prize-eligible drawing list as a CSV file (FR-F5 / US-26). Throws
 * ApiError(404) when nothing is published, like the reports above.
 *
 * Takes the same `challengeId` as the reports, and the dashboard passes it: the
 * export is a record an admin acts on, and a drawing run against last semester's
 * list because the download ignored the selector would be a real error.
 *
 * Fetched rather than linked so the errors land in the app's own empty/alert
 * states instead of navigating the admin to a raw JSON page — and so the
 * session cookie and VITE_API_BASE are handled the same way as every other call.
 */
export async function exportPrizeCsv(
  challengeId?: number,
): Promise<{ blob: Blob; filename: string }> {
  const res = await fetch(`${BASE}/prize-eligible.csv${scope(challengeId)}`, {
    credentials: "include",
  });
  if (!res.ok) await throwApiError(res);
  return {
    blob: await res.blob(),
    filename: filenameFrom(
      res.headers.get("Content-Disposition"),
      "prize-eligible.csv",
    ),
  };
}

File: backend/app/routers/reports.py
from __future__ import annotations

import csv
import io
import re
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy.orm import Session

from app.auth.deps import require_admin
from app.db import get_db
from app.models.challenge import Challenge
from app.schemas.enrollment import (
    NO_ACTIVE_CHALLENGE_CODE,
    NO_ACTIVE_CHALLENGE_MESSAGE,
)
from app.schemas.report import (
    AttendanceReportOut,
    EngagementReportOut,
    ParticipationReportOut,
)
from app.services import challenges as challenge_svc
from app.services import reports as report_svc

router = APIRouter(prefix="/api/reports", tags=["reports"])

PRIZE_CSV_COLUMNS = [
    "student_id",
    "sso_subject",
    "required_completed",
    "required_total",
    "eligible_since",
]


def _report_challenge_or_404(
    db: Session, claims: dict, challenge_id: int | None
) -> Challenge:
    """Resolve the challenge every report in this router describes.

    Omit ``challenge_id`` and this answers for the campus's active challenge —
    "the challenge running right now", which is what an admin opening the
    dashboard means. Pass one and it answers for that challenge instead, which is
    what US-23's "both can be viewed per challenge" asks for and what makes a past
    semester reportable at all.

    Either way the campus comes from the caller's own claims and never from the
    request, so campus isolation is enforced here, before any counting happens.
    An id belonging to another campus is a **404, not a 403** — the convention
    services/challenges.py already follows, so a report cannot be used to probe
    which challenge ids exist elsewhere.

    Published-only in both branches. The active resolver only ever returns a
    published challenge, and an explicit id is held to the same rule rather than
    quietly becoming the one way to report on a draft: "draft" means still being
    authored, and its counts would describe a challenge no student could join.

    Still shared by every route in this file, and now more load-bearing than
    before: it is why one selector moves all four answers at once, so two cards on
    one dashboard can never disagree about which challenge they describe.
    """
    if challenge_id is None:
        challenge = challenge_svc.get_active_challenge_for_campus(db, claims["campus_id"])
    else:
        challenge = challenge_svc.get_challenge(db, claims["campus_id"], challenge_id)
        if challenge is not None and challenge.status != "published":
            challenge = None

    if challenge is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={
                "code": NO_ACTIVE_CHALLENGE_CODE,
                "message": NO_ACTIVE_CHALLENGE_MESSAGE,
            },
        )
    return challenge


def _utc_iso(ts: datetime) -> str:
    """One timestamp shape in the file, whatever the database gave back.

    Check-ins are always written in UTC, but SQLite doesn't keep the offset that
    DateTime(timezone=True) carries, so the same row reads back naive on sqlite
    and UTC-aware on postgres — and would otherwise export two different strings.
    Re-attach the offset the writer guaranteed and print one of them. Seconds
    resolution: this column says which day someone qualified, not which
    microsecond, and the file's own order already carries the tie-break.
    """
    aware = ts.replace(tzinfo=timezone.utc) if ts.tzinfo is None else ts
    return (
        aware.astimezone(timezone.utc)
        .isoformat(timespec="seconds")
        .replace("+00:00", "Z")
    )


def _prize_csv_filename(challenge: Challenge) -> str:
    """A filename an admin can recognise in their downloads folder.

    The semester is admin-entered free text, so anything outside [A-Za-z0-9] is
    collapsed to a dash before it reaches a Content-Disposition header. The
    challenge id keeps two same-semester challenges from colliding.
    """
    slug = re.sub(r"[^A-Za-z0-9]+", "-", challenge.semester).strip("-")
    if not slug:
        return f"prize-eligible-{challenge.id}.csv"
    return f"prize-eligible-{slug}-{challenge.id}.csv"


@router.get("/participation", response_model=ParticipationReportOut)
def participation_report(
    challenge_id: int | None = Query(None),
    claims: dict = Depends(require_admin),
    db: Session = Depends(get_db),
):
    """Participation and the per-week completion funnel (FR-F1 / US-21)."""
    return report_svc.participation_report(
        db, _report_challenge_or_404(db, claims, challenge_id)
    )


@router.get("/attendance", response_model=AttendanceReportOut)
def attendance_report(
    challenge_id: int | None = Query(None),
    claims: dict = Depends(require_admin),
    db: Session = Depends(get_db),
):
    """Auto-vs-manual attendance breakdown (FR-F2 / US-22).

    Its own route rather than a block on /participation: the two count different
    things — students who finished a week vs. check-ins captured — and one grain
    per response keeps each readable as one FR. Both resolve the challenge the
    same way, so they always answer for the same one and 404 together.
    """
    return report_svc.attendance_report(
        db, _report_challenge_or_404(db, claims, challenge_id)
    )


@router.get("/engagement", response_model=EngagementReportOut)
def engagement_report(
    challenge_id: int | None = Query(None),
    claims: dict = Depends(require_admin),
    db: Session = Depends(get_db),
):
    """Content views and guide usage (FR-F3 / US-23).

    Its own route for the reason /attendance is: a third grain — content looked
    at, rather than weeks finished or check-ins captured — and one grain per
    response keeps each response readable as one FR.

    ``challenge_id`` is what US-23's "both can be viewed per challenge" asks for.
    It lands on every route in this file rather than just this one so the cards
    on one dashboard always describe the same challenge.
    """
    return report_svc.engagement_report(
        db, _report_challenge_or_404(db, claims, challenge_id)
    )


@router.get("/prize-eligible.csv", response_class=Response)
def prize_eligible_csv(
    challenge_id: int | None = Query(None),
    claims: dict = Depends(require_admin),
    db: Session = Depends(get_db),
):
    """The prize drawing list as CSV (FR-F5 / US-26).

    Campus-scoped exactly like the reports above, and takes the same
    ``challenge_id``. That it follows the dashboard's selector is not a nicety:
    the export is a *record* an admin acts on, and a drawing run against last
    semester's list because the file quietly ignored the selector would be a real
    error with real prizes attached.

    The header row is written unconditionally, so "nobody is eligible yet" is an
    empty-but-valid CSV rather than an error: it is a true answer, and the admin
    can see the file is the right shape.

    A buffered Response rather than StreamingResponse: one campus cohort is small
    enough to build in memory, and the whole body has to exist anyway to set an
    accurate download.
    """
    challenge = _report_challenge_or_404(db, claims, challenge_id)
    rows = report_svc.prize_eligible_students(db, challenge)

    buf = io.StringIO()
    writer = csv.writer(buf)
    writer.writerow(PRIZE_CSV_COLUMNS)
    for row in rows:
        writer.writerow(
            [
                row.student_id,
                row.sso_subject,
                row.required_completed,
                row.required_total,
                _utc_iso(row.eligible_since),
            ]
        )

    filename = _prize_csv_filename(challenge)
    return Response(
        content=buf.getvalue(),
        media_type="text/csv",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )
